mesonbuild/mtest.py

```diff
@@ -1849,36 +1863,52 @@ def test_suitable(self, test: TestSerialisation) -> bool:
 
         return True
 
-    def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestSerialisation, None, None]:
+    @staticmethod
+    def make_selection_checker(patterns: T.List[str], should_match_if_no_patterns: bool = False) -> T.Callable[[TestSerialisation], bool]:
         '''
-        Allow specifying test names like "meson test foo1 foo2", where test('foo1', ...)
+        Allow specifying test names like "foo1", which will match test('foo1', ...)
 
         Also support specifying the subproject to run tests from like
-        "meson test subproj:" (all tests inside subproj) or "meson test subproj:foo1"
-        to run foo1 inside subproj. Coincidentally also "meson test :foo1" to
+        "subproj:" (all tests inside subproj) or "subproj:foo1"
+        to run foo1 inside subproj. Coincidentally also ":foo1" to
         run all tests with that name across all subprojects, which is
-        identical to "meson test foo1"
+        identical to "foo1"
         '''
-        for arg in self.options.args:
-            if ':' in arg:
-                subproj, name = arg.split(':', maxsplit=1)
+        if not patterns:
+            return lambda _: should_match_if_no_patterns
+
+        unqualified = set()  # type: T.Set[str]
+        qualified = {}  # type: T.Dict[str, T.Union[T.Set[str], UniversalSet[str]]]
```

Comment:
`unqualified: T.Set[str] = set()` and `qualified: T.Dict[str, T.Union[T.Set[str], UniversalSet[str]]] = set()`